README
Little graphics library (Java version).
Copyright (c) 2014-2016, Arnaud Giersch.

This is a little library based on the Java/Swing framework to write simple
programs with graphics in Java.

It is not designed to write "real" programs, but rather to write toy
programs in beginners programming courses.

See the examples and documentation (in French) for details.


CHANGES
-- mar. 4 nov. 2014
        * Correction de la boîte englobante pour {draw,fill}Triangle().

-- mar. 21 oct. 2014
        * Les méthodes *sleep() sont `public'.
        * Vérification des coordonnées dans drawPoint().
        * Correction de getPointColor().

-- ven. 17 oct. 2014
        * Correction d'un décalage avec getMouse{X,Y}.
-- lun. 14 oct. 2014
        * Ajout des méthodes setColor(int) et setBgColor(int).
        * Liste de noms de couleurs plus complète.
        * Ne termine l'application que lorsque la dernière fenêtre est fermée.
        * Ajout des méthodes waitMousePress() et getMouse{X,Y,Button}().


Test.java
class Test{
    public static void main(String[] args) {
        DrawingWindow w1 = new DrawingWindow("Test!", 400, 300);
        final int dy = 25;

        w1.setColor("lawngreen");
        for (int i = 0; i < 12; i++) {
            int p = 10 * i + 10;
            w1.drawLine(p, 0, p, w1.height - 1);
            w1.drawLine(p + i, 0, p + i, w1.height - 1);
        }

        w1.setColor("black");
        for (int i = 0; i < 12; i++) {
            int p = 10 * i + 10;
            int y = 0;

            y += dy;
            w1.drawText(150, y + 5, "circles");
            w1.drawCircle(p, y, i);
            y += dy;
            w1.drawText(150, y + 5, "filled circles");
            w1.fillCircle(p, y, i);

            y += dy;
            w1.drawText(150, y + 10, "rectangles with lines");
            w1.drawLine(p, y, p, y + i);
            w1.drawLine(p, y + i, p + i, y + i);
            w1.drawLine(p + i, y + i, p + i, y);
            w1.drawLine(p + i, y, p, y);

            y += dy;
            w1.drawText(150, y + 10, "rectangles");
            w1.drawRect(p, y, p + i, y + i);
            y += dy;
            w1.drawText(150, y + 10, "filled rectangles");
            w1.fillRect(p, y, p + i, y + i);

            y += dy;
            w1.drawText(150, y + 10, "triangles with lines");
            w1.drawLine(p, y, p + i, y + i/2);
            w1.drawLine(p + i, y + i/2, p, y + i);
            w1.drawLine(p, y + i, p, y);

            y += dy;
            w1.drawText(150, y + 10, "triangles");
            w1.drawTriangle(p, y, p + i, y + i/2, p, y + i);
            y += dy;
            w1.drawText(150, y + 10, "filled triangles");
            w1.fillTriangle(p, y, p + i, y + i/2, p, y + i);

        }

        // Try out of bounds drawings
        w1.setColor("blue");
        w1.drawLine(-10, w1.height - 10, w1.width + 10, w1.height - 10);
        w1.drawLine(w1.width - 10, -10, w1.width - 10, w1.height + 10);
        w1.setColor("red");
        for (int x = -10; x <= w1.width + 10; x++) {
            int y = w1.height - 20;
            w1.drawPoint(x, y);
            int c = w1.getPointColor(x, y);
            if (c != (x < 0 || x >= w1.width ? 0 : 0x00ff0000))
                throw new AssertionError("Error with getPointColor(): " +
                                         "(" + x + ", " + y + "): " +
                                         String.format("%#010x", c));
        }
        for (int y = -10; y <= w1.height + 10; y++) {
            int x = w1.width - 20;
            w1.drawPoint(x, y);
            w1.getPointColor(x, y);
            int c = w1.getPointColor(x, y);
            if (c != (y < 0 || y >= w1.height ? 0 : 0x00ff0000))
                throw new AssertionError("Error with getPointColor(): " +
                                         "(" + x + ", " + y + "): " +
                                         String.format("%#010x", c));
        }

        DrawingWindow w2 = new DrawingWindow("Test!", 800, 600);
        w2.setBgColor("red");
        w2.setColor("blue");
        for (int i = 0; i < 3; i++) {
            w2.clearGraph();
            for (int y = 0; y < w2.height; y++) {
                for (int x = 0; x < w2.width; x++) {
                    w2.drawPoint(x, y);
                }
            }
        }
        w2.setColor("white");
        for (int i = 0; i < 3; i++) {
            w2.clearGraph();
            for (int y = 0; y < w2.height; y++) {
                for (int x = 0; x < w2.width; x++) {
                    w2.drawPoint(x, y);
                }
                w2.sync();
            }
        }
        w2.closeGraph();

        System.out.println("Click anywhere on w1...");

        w1.setColor("black");
        while (w1.waitMousePress(5 * 1000)) {
            int x = w1.getMouseX();
            int y = w1.getMouseY();
            System.out.println("[ " + x + " ; " + y + " ] - " +
                               w1.getMouseButton());
            w1.drawLine(x - 5, y, x + 5, y);
            w1.drawLine(x, y - 5, x, y + 5);
        }

        System.out.println("Done!");
    }
}


Exemple2.java
import java.util.*;

class Exemple2 {

    public static void main(String[] args) {
        // Création de la fenêtre
        DrawingWindow w = new DrawingWindow("Exemple 2", 640, 480);

        int width = Math.min(w.width - 1, w.height - 1) / 2;

        // Affichage de rectangles concentriques, avec un dégradé de
        // couleurs
        for (int z = 0; z <= width; z++) {
            float r, g, b;
            float s = 3.0f * z / width;
            if (z <= width / 3.0f) {
                r = 1.0f - s;
                g = s;
                b = 0.0f;
            } else if (z <= 2.0f * width / 3.0f) {
                s -= 1.0f;
                r = 0.0f;
                g = 1.0f - s;
                b = s;
            } else {
                s -= 2.0f;
                r = s;
                g = 0.0f;
                b = 1.0f - s;
            }
            // On change la couleur de dessin...
            w.setColor(r, g, b);
            // ... et on affiche un rectangle
            w.drawRect(z, z, w.width - 1 - z, w.height - 1 - z);
        }
    }
}


MoreEx/Mandel.java
class Mandel {

    static class Parameters {
        // nombre max d'itérations
        int maxiter;
        // zone d'intérêt par défaut
        double Rmin;
        double Rmax;
        double Imin;
        double Imax;
        // facteur d'échelle
        double Rscale;
        double Iscale;
    }

    static final Parameters initial_parameters;
    static {
        initial_parameters = new Parameters();
        initial_parameters.maxiter = 1000;
        initial_parameters.Rmin    = -2.05;
        initial_parameters.Rmax    = 0.55;
        initial_parameters.Imin    = -1.3;
        initial_parameters.Imax    = 1.3;
        initial_parameters.Rscale  = 0.0;
        initial_parameters.Iscale  = 0.0;
    }

    static double sqr(double x)
    {
        return x * x;
    }

    static int check_point(Parameters p, double cr, double ci)
    {
        double zr2, zi2;
        zi2 = sqr(ci);
        if (sqr(cr + 1) + zi2 < 1.0 / 16.0)
            return p.maxiter;
        double x4 = cr - 1.0 / 4.0;
        double q = sqr(x4) + zi2;
        if (q * (q + x4) < zi2 / 4.0)
            return p.maxiter;
        zr2 = sqr(cr);
        double zr = cr;
        double zi = ci;
        int i;
        for (i = 0 ; i < p.maxiter && zr2 + zi2 < 4 ; i++) {
            zi = 2 * zr * zi + ci;
            zr = zr2 - zi2 + cr;
            zr2 = sqr(zr);
            zi2 = sqr(zi);
        }
        return i;
    }

    static void set_color(DrawingWindow w, Parameters p, int i)
    {
        float rouge, vert, bleu;
        if (i >= p.maxiter) {
            rouge = vert = bleu = 0.0f;
        } else {
            int ii = (p.maxiter - 1 - i) % 96;
            if (ii < 32) {
                // vert -> bleu
                bleu = ii / 32.0f;
                vert = 1.0f - bleu;
                rouge = 0.0f;
            } else if (ii < 64) {
                // bleu -> rouge
                rouge = (ii - 32) / 32.0f;
                bleu = 1.0f - rouge;
                vert = 0.0f;
            } else {
                // rouge -> vert
                vert = (ii - 64) / 32.0f;
                rouge = 1.0f - vert;
                bleu = 0.0f;
            }
        }
        w.setColor(rouge, vert, bleu);
    }

    // Fonction de dessin de l'ensemble de Madelbrot, dans la zone
    // spécifiée, et avec la précision souhgaitée.
    static void do_mandel(DrawingWindow w, Parameters p)
    {
        int x, y;                   // le pixel considéré
        double cr, ci;              // le complexe correspondant
        for (y = 0 ; y < w.height ; y++) {
            ci = p.Imax - y * p.Iscale;
            cr = p.Rmin;
            int x0 = 0;
            int i0 = check_point(p, cr, ci);
            for (x = 1 ; x < w.width ; x++) {
                cr = p.Rmin + x * p.Rscale;
                int i = check_point(p, cr, ci);
                if (i != i0) {
                    set_color(w, p, i0);
                    w.drawLine(x0, y, x - 1, y);
                    i0 = i;
                    x0 = x;
                }
            }
            set_color(w, p, i0);
            w.drawLine(x0, y, w.width - 1, y);
        }
    }

    // Fonction de dessin principale, calcule la zone d'intérêt, appelle
    // do_mandel(), pour dessiner l'ensemble, et permet le zoom.
    static void mandel(DrawingWindow w)
    {
        Parameters p = initial_parameters;
        while (true) {
            p.Rscale = (p.Rmax - p.Rmin) / (w.width - 1);
            p.Iscale = (p.Imax - p.Imin) / (w.height - 1);
            do_mandel(w, p);
            w.setColor("white");
            w.drawText(5, 15, "Cliquer sur l'image pour zoomer");
            w.waitMousePress();
            int x = w.getMouseX();
            int y = w.getMouseY();
            // calcul des coordonnées du point cliqué
            double Tr = p.Rmin + x * p.Rscale;
            double Ti = p.Imax - y * p.Rscale;
            // calcul de la nouvelle zone d'intérêt :
            // zoom ×2 en direction du point cliqué
            final int zoom = 2;
            double Rmin2 = p.Rmin / zoom;
            double Rmax2 = p.Rmax / zoom;
            double Imin2 = p.Imin / zoom;
            double Imax2 = p.Imax / zoom;
            double Rshift = Tr - (Rmin2 + Rmax2) / 2;
            double Ishift = Ti - (Imin2 + Imax2) / 2;
            Rmin2 += Rshift;
            Rmax2 += Rshift;
            Imin2 += Ishift;
            Imax2 += Ishift;
            // affichage d'un rectangle autour de la nouvelle zone d'intérêt
            w.setColor("white");
            w.drawRect((int)Math.round((Rmin2 - p.Rmin) / p.Rscale),
                       (int)Math.round((p.Imax - Imin2) / p.Iscale),
                       (int)Math.round((Rmax2 - p.Rmin) / p.Rscale),
                       (int)Math.round((p.Imax - Imax2) / p.Iscale));
            p.Rmin = Rmin2;
            p.Rmax = Rmax2;
            p.Imin = Imin2;
            p.Imax = Imax2;
        }
    }

    public static void main(String[] args) {
        DrawingWindow w = new DrawingWindow("Mandelbrot' set", 800, 800);
        mandel(w);
    }
}


MoreEx/JeuDeLaVie.java
import java.util.*;

class JeuDeLaVie {

    static final Random random = new Random();

    static final int LARGEUR = 1200;
    static final int HAUTEUR = 900;

    static final int VIVANT = 0x000000ff; // bleu
    static final int MORT   = 0x00ffffff; // blanc

    static class Cell {
        int neigh;
        boolean now;
        boolean next;
    };
    static Cell[][] cells;

    static void majVoisins(int i, int j, int delta)
    {
        for (int k = i - 1 ; k <= i + 1 ; ++k)
            for (int l = j - 1 ; l <= j + 1 ; ++l)
                if (k != i || l != j) {
                    int vi = (k + LARGEUR) % LARGEUR;
                    int vj = (l + HAUTEUR) % HAUTEUR;
                    cells[vi][vj].neigh += delta;
                }
    }

    static void dessine(DrawingWindow w, int i, int j, int couleur)
    {
        w.setColor(couleur);
        w.drawPoint(i, j);
    }

    static void init(DrawingWindow w)
    {
        cells = new Cell[LARGEUR][HAUTEUR];
        for (int i = 0 ; i < LARGEUR ; ++i)
            for (int j = 0 ; j < HAUTEUR ; ++j)
                cells[i][j] = new Cell();
        for (int i = 0 ; i < LARGEUR ; ++i)
            for (int j = 0 ; j < HAUTEUR ; ++j)
                cells[i][j].neigh = 0;
        for (int i = 0 ; i < LARGEUR ; ++i)
            for (int j = 0 ; j < HAUTEUR ; ++j) {
                cells[i][j].now = cells[i][j].next = random.nextBoolean();
                if (cells[i][j].now) {
                    majVoisins(i, j, 1);
                    dessine(w, i, j, VIVANT);
                }
            }
    }

    static void update0(DrawingWindow w)
    {
        for (int i = 0 ; i < LARGEUR ; ++i)
            for (int j = 0 ; j < HAUTEUR ; ++j) {
                switch (cells[i][j].neigh) {
                case 2:
                    // cells[i][j].next = cells[i][j].now; // useless
                    break;
                case 3:
                    cells[i][j].next = true;
                    break;
                default:
                    cells[i][j].next = false;
                    break;
                }
            }
    }
    static void update1(DrawingWindow w)
    {
        for (int i = 0 ; i < LARGEUR ; ++i)
            for (int j = 0 ; j < HAUTEUR ; ++j) {
                if (cells[i][j].now) {
                    if (!cells[i][j].next) {
                        cells[i][j].now = false;
                        majVoisins(i, j, -1);
                        dessine(w, i, j, MORT);
                    }
                } else {
                    if (cells[i][j].next) {
                        cells[i][j].now = true;
                        majVoisins(i, j, 1);
                        dessine(w, i, j, VIVANT);
                    }
                }
            }
    }

    public static void main(String[] args) {
        DrawingWindow w = new DrawingWindow("Jeu de la vie", LARGEUR, HAUTEUR);
        w.setBgColor(MORT);
        w.clearGraph();
        init(w);
        w.sync();
        for (int gen = 0 ; ; ++gen) {
            if (gen % 10 == 0)
                System.err.println("generation " + gen);
            update0(w);
            update1(w);
            w.sync();
        }
    }
}


MoreEx/Chateaux.java
import java.util.*;

class Chateaux {

    static final Scanner input = new Scanner(System.in);
    static final Random random = new Random();

    /* Note : les coordonnées réelles vont de -100 à +100 en abscisse, et
     *  de -10 à +140 en ordonnée
     */
    static final double rXMin = -100.0;
    static final double rXMax = 100.0;
    static final double rYMin = -10.0;
    static final double rYMax = 140.0;
    static final double hauteurMin = 10;
    static final double hauteurMax = 130;
    static final double largeurMin = 40;
    static final double largeurMax = 150;
    static final double ventMax = 30;
    static final double largeurChateau = 8.5;
    static final double hauteurChateau = 7;
    static final double positionChateau1 = -85.0;
    static final double positionChateau2 = 85.0;
    static final double g = 9.81;
    static final double k = 0.005;
    static final double dt = 0.05;
    static int nbJoueurs = 2;
    static int score1 = 0;
    static int score2 = 0;
    static double largeurMont;
    static double hauteurMont;
    static double wnd;

    /* Retourne un nombre pseudo-aléatoire compris entre le paramètre
     * 'min' (inclus) et le paramètre 'max' (exclus)
     */
    static double frand(double min, double max)
    {
        return min + (max - min)* (random.nextDouble());
    }

    // conversion coordonnées réelles -> coordonnées fenêtre
    static int rtowX(DrawingWindow w, double rx)
    {
        return (int)Math.round((w.width - 1) * (rx - rXMin) / (rXMax - rXMin));
    }

    static int rtowY(DrawingWindow w, double ry)
    {
        return (int )Math.round((w.height - 1) * (rYMax - ry) / (rYMax - rYMin));
    }

    // conversion coordonnées fenêtre -> coordonnées réelles
    static double wtorX(DrawingWindow w, int wx)
    {
        return (rXMax - rXMin) * wx / (w.width - 1) + rXMin;
    }

    static double wtorY(DrawingWindow w, int wy)
    {
        return -(rYMax - rYMin) * wy / (w.height - 1) + rYMax;
    }

    static double hauteurMontagne(double largeur, double hauteur, double x)
    {
        double rx = 2.0 * x / largeur;
        return hauteur * (1.0 - rx * rx);
    }

    static void dessineTerrain(DrawingWindow w, double largeur, double hauteur)
    {
        int y0 = rtowY(w, 0) + 1;
        int xmin = rtowX(w, -largeur / 2.0) - 1;
        int xmax = rtowX(w, largeur / 2.0) + 1;
        w.setColor("forestgreen");
        for (int x = xmin; x <= xmax; x++) {
            double rx = wtorX(w, x);
            double ry = hauteurMontagne(largeur, hauteur, rx);
            int y = rtowY(w, ry);
            if (y <= y0)
                w.drawLine(x, y0, x, y);
        }
        w.setColor("maroon");
        w.fillRect(0, y0 + 1, w.width - 1, w.height - 1);
    }

    static void dessineChateau(DrawingWindow w, double position)
    {
        w.setColor("black");
        w.setColor("darkslategray");
        int y1 = rtowY(w, 0) + 1;
        int h0 = rtowY(w, 3);
        int h1 = rtowY(w, 4);
        for (int i = 0; i < 7; i++) {
            int h = i % 2 != 0 ? h0 : h1;
            int x1 = rtowX(w, position + i - 3.5);
            int x2 = rtowX(w, position + i - 2.5) - 1;
            w.fillRect(x1, y1, x2, h);
        }
        w.setColor("dimgray");
        h0 = rtowY(w, 6);
        h1 = rtowY(w, 7);
        for (int i = 0; i < 5; i++) {
            int h = i % 2 != 0 ? h0 : h1;
            int x1 = rtowX(w, position + i - 8.5);
            int x2 = rtowX(w, position + i - 7.5) - 1;
            w.fillRect(x1, y1, x2, h);
            x1 = rtowX(w, position + i + 3.5);
            x2 = rtowX(w, position + i + 4.5) - 1;
            w.fillRect(x1, y1, x2, h);
        }
    }

    static void dessineVent(DrawingWindow w, double vitesse)
    {
        int lg = rtowX(w, vitesse) - rtowX(w, 0);
        int dir = lg > 0 ? 1 : -1;
        int y = 20;
        w.setColor("black");
        if (lg == 0) {
            w.drawCircle(w.width / 2, y, 4);
        } else {
            int x1 = (w.width - lg) / 2;
            int x2 = (w.width + lg) / 2;
            w.drawLine(x1 - dir, y - 1, x2 - dir, y - 1);
            w.drawLine(x1, y, x2, y);
            w.drawLine(x1 - dir, y + 1, x2 - dir, y + 1);
            for (int i = 0; i < 3; i++) {
                w.drawLine(x2 - i * dir, y, x2 - (6 + i) * dir, y - 4);
                w.drawLine(x2 - i * dir, y, x2 - (6 + i) * dir, y + 4);
            }
        }
    }

    static void dessineExplosion(DrawingWindow w, double rx, double ry)
    {
        final int maxray = rtowX(w, 2.5) - rtowX(w, 0);
        // 1/2 rouge -> rouge -> jaune
        final int x = rtowX(w, rx);
        final int y = rtowY(w, ry);
        int i;
        for (i = 0; i <= maxray / 3; i++) {
            w.setColor(0.5f + 3.0f * i / (2.0f * maxray), 0.0f, 0.0f);
            w.drawCircle(x, y, i);
            DrawingWindow.msleep(20);
        }
        for (/* i */; i < maxray; i++) {
            w.setColor(1.0f, 1.5f * i / maxray - 0.5f, 0.0f);
            w.drawCircle(x, y, i);
            DrawingWindow.msleep(20);
        }
        w.setColor("skyblue");
        for (i = 0; i < maxray; i++) {
            w.drawCircle(x, y, i);
            DrawingWindow.msleep(10);
        }
        //    w.fillCircle(x, y, maxray - 1);
    }

    static void dessineFlammes(DrawingWindow w, double x0, double y0)
    {
        for (int i = 0; i < 70; i++) {
            double dt = 0.05;
            double vx = frand(-2.5, 2.5);
            double vy = frand(5, 17);
            double x = x0;
            double y = y0;
            float red = (float)frand(0.5, 1);
            float green = (float)frand(0, red);
            float blue = 0;
            w.setColor(red, green, blue);
            while (y >= 0.0) {
                w.drawPoint(rtowX(w, x), rtowY(w, y));
                x += vx * dt;
                y += vy * dt;
                vy -= 9.81 * dt;
            }
            DrawingWindow.msleep(30);
        }
    }

    static void initialise(DrawingWindow w)
    {
        largeurMont = frand(largeurMin, largeurMax);
        hauteurMont = frand(hauteurMin, hauteurMax);
        wnd = frand(-ventMax, ventMax);
        w.setBgColor("skyblue");
        w.clearGraph();
        dessineTerrain(w, largeurMont, hauteurMont);
        dessineVent(w, wnd);
        dessineChateau(w, positionChateau1);
        dessineChateau(w, positionChateau2);
        w.setColor("wheat");
        w.drawText(rtowX(w, positionChateau1 - largeurChateau), rtowY(w, 0) + 20,
                   "Joueur 1");
        w.drawText(rtowX(w, positionChateau2 - largeurChateau), rtowY(w, 0) + 20,
                   "Joueur 2");
        w.drawText(rtowX(w, 0) - 15, rtowY(w, 0) + 20, score1 + " / " + score2);
    }

    /* Retour : numéro du perdant, 0 sinon
       x et y contiennent les coordonnées de la collision
    */
    static class ResTir {
        int perdant;
        double x;
        double y;
    }
    static ResTir tir(DrawingWindow w,
                      double x0, double y0, double v0, double alpha) {
        double vx = v0 * Math.cos(alpha);
        double vy = v0 * Math.sin(alpha);
        double x = x0;
        double y = y0;
        int collision = 0;
        do {
            int wx = rtowX(w, x);
            int wy = rtowY(w, y);
            w.setColor("black");
            w.fillCircle(wx, wy, 2);
            double vxr = vx - wnd;
            double kvr = -k * Math.sqrt(vxr * vxr + vy * vy);
            double ax = kvr * vxr;
            double ay = kvr * vy - g;
            x += vx * dt;
            y += vy * dt;
            vx += ax * dt;
            vy += ay * dt;

            DrawingWindow.msleep(10);
//             w.sync();
            w.setColor("skyblue");
            w.fillCircle(wx, wy, 2);
//             w.setColor("black");
//             w.drawPoint(wx, wy);
            if (y <= 0) {
                collision = 3;
            } else if (y < hauteurChateau) {
                if (positionChateau1 - largeurChateau <= x
                    && positionChateau1 + largeurChateau >= x)
                    collision = 1;
                else if (positionChateau2 - largeurChateau <= x
                         && positionChateau2 + largeurChateau >= x)
                    collision = 2;
            }
            if (collision == 0) {
                double h = hauteurMontagne(largeurMont, hauteurMont, x);
                if (h > 0 && y < h)
                    collision = 3;
            }
        } while (collision == 0);
        ResTir res = new ResTir();
        res.x = x;
        res.y = y;
        res.perdant = collision == 3 ? 0 : collision;
        return res;
    }

    static int jeu1(DrawingWindow w)
    {
        initialise(w);
        int joueur = 2;
        double x, y;
        int perdant;
        do {
            joueur = 3 - joueur;
            w.sync();
            System.out.println("-=| Joueur " + joueur + " |=-");
            double alpha;
            double v0;
            if (joueur <= nbJoueurs) {
                System.out.print("\nangle ? ");
                alpha = input.nextDouble();
                System.out.print("vitesse initiale ? ");
                v0 = input.nextDouble();
            } else {
                alpha = frand(10, 90);
                v0 = frand(10, 100);
                System.out.println(" [ " + (int )alpha + "° ; " + (int )v0 + " ]");
            }
            alpha = Math.toRadians(alpha);
            double x0;
            if (joueur == 1) {
                x0 = positionChateau1 + 0.8 * largeurChateau;
            } else {
                x0 = positionChateau2 - 0.8 * largeurChateau;
                alpha = Math.PI - alpha;
            }
            double y0 = hauteurChateau + 1;
            {
                ResTir r = tir(w, x0, y0, v0, alpha);
                x = r.x;
                y = r.y;
                perdant = r.perdant;
            }
            dessineExplosion(w, x, y);
            dessineVent(w, wnd);
        } while (perdant == 0);
        dessineFlammes(w, x, y);
        String msg = " Joueur " + perdant;
        if (perdant == joueur)
            msg += " s'est suicidé ! ";
        else
            msg += " a perdu ! ";
        w.setColor("darkred");
        w.setBgColor("white");
        w.drawText(w.width / 2, w.height / 3, msg);
        w.sync();
        System.out.println(msg);
        return perdant;
    }

    public static void main(String[] args) {
        DrawingWindow w = new DrawingWindow("Chateaux", 640, 480);

        if (args.length > 0)
            nbJoueurs = Integer.parseInt(args[0]);

        boolean rejouer = true;
        do {
            int perdant = jeu1(w);
            if (perdant == 1)
                score2++;
            else if (perdant == 2)
                score1++;
            System.out.println("### SCORE : " + score1 + " / " + score2 + " ###");
            if (nbJoueurs == 0)
                DrawingWindow.sleep(2);
            else {
                char r;
                do {
                    System.out.println("Recommencer (o/n) ? ");
                    r = input.next().charAt(0);
                } while (r != 'o' && r != 'n');
                rejouer = r == 'o';
            }
        } while (rejouer);
        w.closeGraph();
    }
}
